pkg/controller/resourceclaim/controller.go

```diff
@@ -314,82 +334,204 @@ func (ec *Controller) syncPod(ctx context.Context, namespace, name string) error
+			if status.SourceRef.Name == nil || newPodClaims[*status.SourceRef.Name] == "" {
+				updatedPod.Status.ResourceClaimStatuses = append(updatedPod.Status.ResourceClaimStatuses, status)
+			}
+		}
+
+		// Sorting is not strictly necessary, but nicer.
+		sort.Slice(updatedPod.Status.ResourceClaimStatuses, func(i, j int) bool {
+			return pointer.StringDeref(updatedPod.Status.ResourceClaimStatuses[i].SourceRef.Name, "") <
+				pointer.StringDeref(updatedPod.Status.ResourceClaimStatuses[j].SourceRef.Name, "")
+		})
+
+		// No one else is currently setting this field, so there's no
+		// reason to worry about overwriting information stored by
+		// someone else.
+		//
+		// TODO: tests...
+		if _, err := ec.kubeClient.CoreV1().Pods(namespace).UpdateStatus(ctx, updatedPod, metav1.UpdateOptions{}); err != nil {
+			return fmt.Errorf("update pod %s/%s ResourceClaimStatuses: %v", namespace, name, err)
+		}
+	}
+
 	return nil
 }
 
 // handleResourceClaim is invoked for each volume of a pod.
-func (ec *Controller) handleClaim(ctx context.Context, pod *v1.Pod, podClaim v1.PodResourceClaim) error {
+func (ec *Controller) handleClaim(ctx context.Context, pod *v1.Pod, podClaim v1.PodResourceClaim, newPodClaims *map[string]string) error {
 	logger := klog.LoggerWithValues(klog.FromContext(ctx), "podClaim", podClaim.Name)
 	ctx = klog.NewContext(ctx, logger)
 	logger.V(5).Info("checking", "podClaim", podClaim.Name)
+
+	// resourceclaim.Name checks for the situation that the client doesn't
+	// know some future addition to the API. Therefore it gets called here
+	// even if there is no template to work on, because if some new field
+	// gets added, the expectation might be that the controller does
+	// something for it.
+	claimName, mustCheckOwner, err := resourceclaim.Name(pod, &podClaim)
+	switch {
+	case errors.Is(err, resourceclaim.ErrClaimNotFound):
+		// Continue below.
+	case err != nil:
+		return fmt.Errorf("checking for claim before creating it: %v", err)
+	case claimName == nil:
+		// Nothing to do, no claim needed.
+	case *claimName != "":
+		claimName := *claimName
+		// The ResourceClaim should exist because it is recorded in the pod.status.resourceClaimStatuses,
+		// but perhaps it was deleted accidentally. In that case we re-create it.
+		claim, err := ec.claimLister.ResourceClaims(pod.Namespace).Get(claimName)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		if claim != nil {
+			var err error
+			if mustCheckOwner {
+				err = resourceclaim.IsForPod(pod, claim)
+			}
+			if err == nil {
+				// Already created, nothing more to do.
+				logger.V(5).Info("claim already created", "podClaim", podClaim.Name, "resourceClaim", claimName)
+				return nil
+			}
+			logger.Error(err, "claim that was created for the pod is no longer owned by the pod, creating a new one", "podClaim", podClaim.Name, "resourceClaim", claimName)
+		}
+	}
+
 	templateName := podClaim.Source.ResourceClaimTemplateName
 	if templateName == nil {
```

Comment:
This is where previously, the controller would have returned without an error when `podClaim.Source` was empty (= future unknown field set).

Treating this as an error (new code) is consistent with the API description and while it may be too strict, it's probably safer because if a new field gets added that has to be handled by the kube-controller-manager, then running with an older release will fail, as it should in this case.